docs/site/Context.md

```diff
@@ -530,3 +530,98 @@ class MyController {
   }
 }
 ```
+
+## Configuration by convention
+
+To allow bound items in the context to be configured, we introduce some
+conventions and corresponding APIs to make it simple and consistent.
+
+We treat configurations for bound items in the context as dependencies, which
+can be resolved and injected in the same way of other forms of dependencies. For
+example, the `RestServer` can be configured with `RestServerConfig`.
+
+Let's first look at an example:
+
+```ts
+export class RestServer {
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE) app: Application,
+    @inject(RestBindings.CONFIG, {optional: true})
+    config: RestServerConfig = {},
+  ) {
+    // ...
+  }
+  // ...
+}
+```
+
+The configuration itself is a binding (`RestBindings.CONFIG`) in the context.
+It's independent of the binding for `RestServer`. The caveat is that we need to
+maintain a different binding key for the configuration. Referencing a hard-coded
+key for the configuration also makes it impossible to have more than one
+instances of the `RestServer` to be configured with different options, such as
+`protocol` or `port`.
+
+To solve these problems, we introduce an accompanying binding for an item that
+expects configuration. For example:
+
+- `servers.RestServer.server1`: RestServer
+- `servers.RestServer.server1.$config`: RestServerConfig
```

Comment:
Why not simply `servers.RestServer.server1.config`?

Reply:
We use `$config` as a special suffix to avoid possible conflict.

Reply:
Doesn't feel consistent with the rest of the API - the only place where you access the value with a key prefixed with a `$`. Also, technically it doesn't prevent conflict.

What could possibly set a `config` value on `server1`, apart from its own configuration? Maybe `config` should be implemented as a getter-setter?

Reply:
@raymondfeng @bajtos thoughts on the comment above?

Reply:
My understanding:

The idea is to treat `$` as a reserved character that should not be used by user-defined bindings, only by the framework itself. 

- `servers.RestServer.server1.config` is regular binding key that anybody can define and bind values to.
- `servers.RestServer.server1.$config` is an internal key used by the framework only. App & extension developers should never use this key directly, but use helpers like `@config` and `ctx.configure` instead.